### PR TITLE
Add `AsyncFunction` types

### LINKED ISSUE #121: Proposal: `AsyncFunction`
To make it easier to constrain a type to be an async function. We can currently do `T extends Function`, but there's no way to enforce it to be async.

We already have an internal type for this:

https://github.com/sindresorhus/type-fest/blob/78019f42ea888b0cdceb41a4a78163868de57555/source/async-return-type.d.ts#L3

But I'm thinking it could be more reusable:

```ts
type AsyncFunction<Arguments extends readonly any[] = readonly any[], ReturnValue = unknown> = (...arguments_: Arguments) => Promise<ReturnValue>;
```

Thoughts?

### PR DESCRIPTION
**Pull request by zfaustk**
Closes #121

This adds the three connected Basic function types from the final issue specification:

- `AsyncFunction<Arguments, ReturnValue>` for a reusable asynchronous signature;
- `AnyAsyncFunction` for an erased asynchronous function constraint;
- `AnyFunction` for an erased sync-or-async function constraint.

The definitions follow the names, defaults, and `any` edge-case decision discussed in #121. They live alongside the existing class and constructor types in `source/basic.d.ts`, with README entries and examples.

Tests cover generic arguments and return values, defaults, `any`, rejection of synchronous functions by `AnyAsyncFunction`, acceptance of synchronous and asynchronous functions by `AnyFunction`, and a `never` parameter.

Validation:

- `npx tsd --files test-d/async-function.ts`
- `npm run test:tsd`
- `npm run test:tsc`
- `npm run test:xo`
- `npm run test:linter`

Codex assisted with checking the type edge cases, naming, documentation, examples, and test coverage against the upstream discussion and repository conventions.

### CODE REVIEW

File: source/basic.d.ts [hunk 1]
```diff
@@ -35,4 +35,72 @@ Matches an [`abstract class`](https://www.typescriptlang.org/docs/handbook/relea
...
+
+function identity<FunctionType extends AnyFunction>(function_: FunctionType): FunctionType {
+	return function_;
+}
+```
+
+@category Basic
+*/
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type AnyFunction = (...arguments_: readonly any[]) => any;
```
Comment (sindresorhus) [inline source/basic.d.ts:104]:
```suggestion
export type AnyFunction = (...arguments_: readonly any[]) => unknown;
```

File: test-d/async-function.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+expectAssignable<AnyAsyncFunction>(async (value: string) => value.length);
+expectNotAssignable<AnyAsyncFunction>(async (value: never) => value);
+declare const promiseReturningFunction: (value: {id: string}) => Promise<string>;
+declare const synchronousFunction: (value: {id: string}) => string;
+expectAssignable<AnyAsyncFunction>(promiseReturningFunction);
+expectNotAssignable<AnyAsyncFunction>(synchronousFunction);
+
+expectAssignable<AnyFunction>((value: never) => value);
+expectAssignable<AnyFunction>(async () => 'value');
+expectAssignable<AnyFunction>(() => 'value');
```
Comment (sindresorhus) [inline test-d/async-function.ts:22]: I would add a few more tests:

```ts
// `Arguments` is enforced.
expectError(asyncFunction());
expectError(asyncFunction(123));

// `ReturnValue` is enforced.
expectNotAssignable<AsyncFunction<[value: string], number>>(
	async (value: string) => value,
);

// Defaults accept arbitrary arguments.
expectType<Promise<unknown>>(defaultAsyncFunction('value', 123));

// Erased functions remain callable with an unknown result.
declare const anyFunction: AnyFunction;
expectType<unknown>(anyFunction('value', 123));

declare const erasedAsyncFunction: AnyAsyncFunction;
expectType<Promise<unknown>>(erasedAsyncFunction('value', 123));

// Ordinary required parameters are supported.
expectAssignable<AnyFunction>((value: string) => value);
```

File: source/basic.d.ts (hunk 1 above)
Comment (zfaustk) [reply source/basic.d.ts]: Updated `AnyFunction` to return `unknown` as suggested.

File: test-d/async-function.ts (hunk 2 above)
Comment (zfaustk) [reply test-d/async-function.ts:40]: Added these cases, including direct invalid-call checks for the argument constraints. The full `npm test` suite passes locally.
